src/indexing.jl

```diff
@@ -139,8 +141,9 @@ You can also work with it directly, say you wanted to get all checkpoints files
 2: https://github.com/JuliaData/DataFrames.jl
 """
 function index_checkpoint_files(dir::AbstractPath)
+    isdir(dir) || throw(DomainError(dir, "Need an existing directory."))
```

Comment:
Just to note, I didn't know that S3 `walkpath` does _not_ already error if this is the case, but local `walkpath` does:

```julia
julia> using AWSS3, FilePathsBase

julia> collect(walkpath(p"s3://<some-existing-bucket>/foo/"))
S3Path{Nothing}[]

julia> collect(walkpath(p"foo"))
ERROR: TaskFailedException
Stacktrace:
  [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
    @ Base ./task.jl:710
  [2] wait
    @ ./task.jl:769 [inlined]
  [3] wait(c::Base.GenericCondition{ReentrantLock})
    @ Base ./condition.jl:106
  [4] take_unbuffered(c::Channel{PosixPath})
    @ Base ./channels.jl:405
  [5] take!
    @ ./channels.jl:383 [inlined]
  [6] iterate(c::Channel{PosixPath}, state::Nothing)
    @ Base ./channels.jl:465
  [7] iterate
    @ ./channels.jl:464 [inlined]
  [8] _collect(cont::UnitRange{Int64}, itr::Channel{PosixPath}, #unused#::Base.HasEltype, isz::Base.SizeUnknown)
    @ Base ./array.jl:613
  [9] collect(itr::Channel{PosixPath})
    @ Base ./array.jl:602
 [10] top-level scope
    @ REPL[17]:1

    nested task error: IOError: readdir("foo"): no such file or directory (ENOENT)
    Stacktrace:
     [1] uv_error
       @ ./libuv.jl:97 [inlined]
     [2] readdir(dir::String; join::Bool, sort::Bool)
       @ Base.Filesystem ./file.jl:830
     [3] readdir
       @ ./file.jl:825 [inlined]
     [4] readdir
       @ ~/.julia/packages/FilePathsBase/O5N9l/src/system.jl:386 [inlined]
     [5] readpath(p::PosixPath)
       @ FilePathsBase ~/.julia/packages/FilePathsBase/O5N9l/src/path.jl:701
     [6] (::FilePathsBase.var"#25#26"{Bool, Bool, typeof(throw), PosixPath})(chnl::Channel{PosixPath})
       @ FilePathsBase ~/.julia/packages/FilePathsBase/O5N9l/src/path.jl:711
     [7] (::Base.var"#517#518"{FilePathsBase.var"#25#26"{Bool, Bool, typeof(throw), PosixPath}, Channel{PosixPath}})()
       @ Base ./channels.jl:132
```

Reply:
Perhaps throwing OError would be better, then?

Reply:
I think `ArgumentError` would be suitable here because it's a problem with the `dir` argument.

Reply:
This will throw a different error on S3 vs local paths